### ISSUE TITLE
Spurious  shorthand_optional_binding error when a local variable has the same name as an external variable

### ISSUE THREAD

**Issue by rgoldberg** [realm/SwiftLint#6062]
### New Issue Checklist

- [x] I've Updated SwiftLint to the latest version.
- [x] I've searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues).

### Bug Description

shorthand_optional_binding outputs the following spurious error when a local variable has the same name as an external variable:

```text
Shorthand Optional Binding Violation: Use shorthand syntax for optional binding (shorthand_optional_binding)
```

For the following code:

```swift
var y: String {
    if let x = x { // error for this line
        "A"
    } else {
        "B"
    }
}

private var x: String? {
    "C"
}
```

Mention the command or other SwiftLint integration method that caused the issue. Include stack traces or command output.

```bash
$ swiftlint --strict --quiet <file name>
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure) 0.59.1
* Xcode version (run `xcodebuild -version` to be sure) Xcode 16.3 (16E140)
* Installation method used (Homebrew, CocoaPods, building from source, etc) Homebrew
* Configuration file:

```yml
---
opt_in_rules:
  - all
disabled_rules:
  # eventually enable
  - file_header
  - one_declaration_per_file
  - trailing_comma
  # never enable
  - contrasted_opening_brace
  - explicit_acl
  - explicit_enum_raw_value
  - explicit_top_level_acl
  - explicit_type_interface
  - no_extension_access_modifier
  - no_grouping_extension
  - no_magic_numbers
  - prefixed_toplevel_constant
  - sorted_enum_cases
  - vertical_whitespace_between_cases
closure_body_length:
  warning: 40
file_name:
  excluded: [Finder.swift, SpotlightInstalledApps.swift, Utilities.swift]
file_types_order:
  order:
    [
      [main_type],
      [supporting_type],
      [extension],
      [preview_provider],
      [library_content_provider],
    ]
function_body_length:
  warning: 55
number_separator:
  minimum_length: 6
opening_brace:
  ignore_multiline_statement_conditions: true
type_contents_order:
  order:
    [
      [case],
      [type_alias, associated_type],
      [subtype],
      [type_property],
      [instance_property],
      [ib_inspectable],
      [ib_outlet],
      [initializer],
      [deinitializer],
      [type_method],
      [view_life_cycle_method],
      [ib_action],
      [other_method],
      [subscript],
    ]
```

Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)? If so, paste their
relative paths and respective contents.
